panic: binlog_streamer.go:181] stream ended

#### Overview of the Issue

When starting up a tablet with SHA ae7b400 we see the following:

```
I0410 01:15:26.662889       8 binlog_streamer.go:181] stream ended @ <nil>, err = <nil>
panic: runtime error: slice bounds out of range

goroutine 515 [running]:
vitess.io/vitess/go/mysql.printJSONString(0xc001546c3b, 0xb6, 0x1bb, 0x0, 0xc00104bc70)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_json.go:353 +0x17e
vitess.io/vitess/go/mysql.printJSONValue(0x94110c, 0xc001546c3b, 0xb6, 0x1bb, 0x12c5800, 0xc00104bc70, 0x0, 0x0)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_json.go:98 +0x562
vitess.io/vitess/go/mysql.printJSONValueEntry(0xc001546ae9, 0x208, 0x30d, 0x57, 0x12c5800, 0xc00104bc70, 0x0, 0x0)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_json.go:215 +0x2dc
vitess.io/vitess/go/mysql.printJSONObject(0xc001546ae9, 0x208, 0x30d, 0xc000e8ca00, 0xc00104bc70, 0x40e1c8, 0x70)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_json.go:139 +0x430
vitess.io/vitess/go/mysql.printJSONValue(0x1071500, 0xc001546ae9, 0x208, 0x30d, 0x1, 0xc00104bc70, 0x1, 0x117625c)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_json.go:74 +0x126
vitess.io/vitess/go/mysql.printJSONData(0xc001546ae8, 0x209, 0x30e, 0x0, 0xa, 0xc0010476b0, 0x1, 0x8)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_json.go:65 +0x93
vitess.io/vitess/go/mysql.CellValue(0xc001546aa5, 0x34d, 0x351, 0x3f, 0x81e000400f5, 0x1, 0xc0010476b0, 0x1, 0x8, 0x1, ...)
vitess-build/src/vitess.io/vitess/go/mysql/binlog_event_rbr.go:811 +0xb3d
vitess.io/vitess/go/vt/binlog.writeValuesAsSQL(0xc0015cacc0, 0xc000d97c80, 0xc000e8d9a0, 0x0, 0x0, 0x1, 0xc0012489d0, 0xd, 0x113, 0xc00010814d, ...)
vitess-build/src/vitess.io/vitess/go/vt/binlog/binlog_streamer.go:752 +0x598
vitess.io/vitess/go/vt/binlog.(*Streamer).appendInserts(0xc000dbd860, 0xc001d1c000, 0x1, 0xa, 0xc000d97c80, 0xc000e8d9a0, 0x26, 0x2b, 0xc000dcdc00)
vitess-build/src/vitess.io/vitess/go/vt/binlog/binlog_streamer.go:610 +0x1f8
vitess.io/vitess/go/vt/binlog.(*Streamer).parseEvents(0xc000dbd860, 0x12d9300, 0xc000d7c140, 0xc00024f500, 0xc000e240c0, 0xc00024f500, 0x0, 0x0)
vitess-build/src/vitess.io/vitess/go/vt/binlog/binlog_streamer.go:534 +0x19b0
vitess.io/vitess/go/vt/binlog.(*Streamer).Stream(0xc000dbd860, 0x12d9300, 0xc000d7c140, 0x0, 0x0)
vitess-build/src/vitess.io/vitess/go/vt/binlog/binlog_streamer.go:233 +0x201
vitess.io/vitess/go/vt/vttablet/tabletserver.(*ReplicationWatcher).Process(0xc0002367c0, 0x12d9300, 0xc000d7c140, 0x1d5e800)
vitess-build/src/vitess.io/vitess/go/vt/vttablet/tabletserver/replication_watcher.go:138 +0x1ac
created by vitess.io/vitess/go/vt/vttablet/tabletserver.(*ReplicationWatcher).Open
vitess-build/src/vitess.io/vitess/go/vt/vttablet/tabletserver/replication_watcher.go:97 +0xcd
```

I'll post a schema + sample data when I've got a clear and minimal reproduction.

--- a/go/mysql/binlog_event_json.go
+++ b/go/mysql/binlog_event_json.go
@@ -484,15 +484,17 @@ func readOffsetOrSize(data []byte, pos int, large bool) (int, int) {
 }
 
 func readVariableInt(data []byte, pos int) (int, int) {
-	var b int8
-	var result int
+	var bb byte
+	var res int
+	var idx byte
 	for {
-		b = int8(data[pos])
+		bb = data[pos]
 		pos++
-		result = (result << 7) + int(b&0x7f)
-		if b >= 0 {
+		res |= int(bb&0x7f) << (7 * idx)
+		if int8(bb) >= 0 {
 			break
 		}
+		idx++
 	}
-	return result, pos
+	return res, pos
 }

--- a/go/mysql/binlog_event_json_test.go
+++ b/go/mysql/binlog_event_json_test.go
@@ -47,9 +47,12 @@ func TestJSON(t *testing.T) {
 		data:     []byte{12, 13, 115, 99, 97, 108, 97, 114, 32, 115, 116, 114, 105, 110, 103},
 		expected: `'"scalar string"'`,
 	}, {
+		data:     []byte{0, 1, 0, 149, 0, 11, 0, 6, 0, 12, 17, 0, 115, 99, 111, 112, 101, 115, 130, 1, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 66, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 66, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 65, 69, 65, 65, 65, 65, 65, 65, 69, 65, 65, 65, 65, 65, 65, 56, 65, 65, 65, 66, 103, 65, 65, 65, 65, 65, 65, 66, 65, 65, 65, 65, 67, 65, 65, 65, 65, 65, 65, 65, 65, 65, 84, 216, 142, 184},
+		expected: `JSON_OBJECT('scopes','AAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAEAAAAAAEAAAAAA8AAABgAAAAAABAAAACAAAAAAAAA')`,
+	}, {
 		// repeat the same string 10 times, to test readVariableInt when length of string
 		// requires 2 bytes to store
-		data: []byte{12, 129, 2,
+		data: []byte{12, 130, 1,
 			115, 99, 97, 108, 97, 114, 32, 115, 116, 114, 105, 110, 103,
 			115, 99, 97, 108, 97, 114, 32, 115, 116, 114, 105, 110, 103,
 			115, 99, 97, 108, 97, 114, 32, 115, 116, 114, 105, 110, 103,
@@ -150,7 +153,7 @@ func TestJSON(t *testing.T) {
 				}
 			} else {
 				if got := string(r); got != tcase.expected {
-					t.Errorf("unexpected output for %v: got [%v] expected [%v]", tcase.data, got, tcase.expected)
+					t.Errorf("unexpected output for %v: got \n[%v] \n expected \n[%v]", tcase.data, got, tcase.expected)
 				}
 			}
 

